File: easy/maximum-depth-of-n-ary-tree/maximum-depth-of-n-ary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function (root) {
  if (!root) return 0
  let ret = 1;
  const walk = (node, depth) => {
    if (node) {
      if (node.left) {
        walk(node.left, depth + 1)
      }
      if (node.right) {
        walk(node.right, depth + 1)
      }
    }
    if (depth > ret) ret = depth
  }
  walk(root, 1)
  return ret
};

File: easy/maximum-depth-of-n-ary-tree/maximum-depth-of-n-ary-tree.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""


class Solution:
    def maxDepth(self, root: 'Node') -> int:
        if not root:
            return 0

        def bfs(node, d=1):
            if not node or not node.children:
                return d
            return max([bfs(n, d+1) for n in node.children])
        return bfs(root)


class Solution2(object):
    def maxDepth(self, root):
        if root == None:
            return 0
        depth = 0
        stack = [root]
        while stack:
            next_level = []
            while stack:
                node = stack.pop()
                if node.children:
                    next_level += node.children
            stack = next_level
            depth += 1
        return depth
